OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,302 @@
+    service_state_.exit_requested = true;
+  }
+
+  LOG("Attempting to disconnect from CaptureEventProducers as exit was requested");
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    for (grpc::ServerContext* context : server_contexts_) {
+      // This should cause blocking Reads on ServerReaderWriter to fail immediately.
+      context->TryCancel();
+    }
+  }
+
+  {
+    absl::WriterMutexLock lock{&capture_event_buffer_mutex_};
+    capture_event_buffer_ = nullptr;
+  }
+}
+
+grpc::Status ProducerSideServiceImpl::ReceiveCommandsAndSendEvents(
+    ::grpc::ServerContext* context,
+    ::grpc::ServerReaderWriter< ::orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                                ::orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream) {
+  LOG("A CaptureEventProducer has connected calling ReceiveCommandsAndSendEvents");
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.emplace(context);
+  }
+
+  // This will also be protected by service_state_mutex_.
+  bool all_events_sent_received = true;
+
+  std::atomic<bool> exit_send_commands_thread = false;
+
+  std::thread send_commands_thread{&ProducerSideServiceImpl::SendCommandsThread,
+                                   this,
+                                   context,
+                                   stream,
+                                   &all_events_sent_received,
+                                   &exit_send_commands_thread};
+  std::thread receive_events_thread{&ProducerSideServiceImpl::ReceiveEventsThread, this, context,
+                                    stream, &all_events_sent_received};
+
+  receive_events_thread.join();
+  // When receive_events_thread exits because stream->Read(&request) fails,
+  // it means that the producer has disconnected: ask send_commands_thread to exit, too.
+  exit_send_commands_thread = true;
+  send_commands_thread.join();
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.erase(context);
+  }
+
+  LOG("Finished handling ReceiveCommandsAndSendEvents for a CaptureEventProducer");
+  return grpc::Status::OK;
+}
+
+void ProducerSideServiceImpl::SendCommandsThread(
+    grpc::ServerContext* context,
+    grpc::ServerReaderWriter<orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                             orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream,
+    bool* all_events_sent_received, std::atomic<bool>* exit_send_commands_thread) {
+  // As a result of initializing prev_capture_status to kCaptureFinished,
+  // an initial StartCaptureCommand is sent
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStarted,
+  // and an initial StopCaptureCommand is sent (with little effect)
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStopping.
+  CaptureStatus prev_capture_status = CaptureStatus::kCaptureFinished;
+
+  while (true) {
```

Comment:
A general description of that loop (or just the whole method), would be beneficial.